src/main/java/com/endava/internship/collections/StudentSet.java

```diff
@@ -1,84 +1,250 @@
-        //TODO
-        return false;
-    }
+        public StudentSet(){
+            list = new LinkedList<>();
+            root = new Node(null); //??
+        }
 
-    @Override
-    public boolean contains(Object o) {
-        //TODO
-        return false;
-    }
+        @Override
+        public int size() {
+            return size;
+        }
 
-    @Override
-    public Iterator<Student> iterator() {
-        //TODO
-        return null;
-    }
+        @Override
+        public boolean isEmpty() {
+            return size == 0;
+        }
 
-    @Override
-    public Object[] toArray() {
-        //TODO
-        return new Object[0];
-    }
+        @Override
+        public boolean contains(Object o) {
+            if (!(o instanceof Student)){
+                return false;
+            }
 
-    @Override
-    public <T> T[] toArray(T[] ts) {
-        //TODO
-        return null;
-    }
+            Node eNode = new Node((Student) o);
+            return search(root, eNode) != null;
+        }
 
-    @Override
-    public boolean add(Student student) {
-        //TODO
-        return false;
-    }
+        private Node search(Node node, Node eNode){
+            int compare = node.compareTo(eNode);
 
-    @Override
-    public boolean remove(Object o) {
-        //TODO
-        return false;
-    }
+            if (compare < 0 && node.right != null){
+                return search(node.right, eNode);
+            }
 
-    @Override
-    public void clear() {
-        //TODO
-    }
+            if (compare > 0 && node.left != null){
+                return search(node.left, eNode);
+            }
 
-    @Override
-    public boolean addAll(Collection<? extends Student> collection) {
-        //TODO
-        return false;
-    }
+            if (compare == 0){
+                return node;
+            }
 
-    @Override
-    public boolean containsAll(Collection<?> collection) {
-        //Ignore this for homework
-        throw new UnsupportedOperationException();
-    }
+            return null;
+        }
 
-    @Override
-    public boolean retainAll(Collection<?> collection) {
-        //Ignore this for homework
-        throw new UnsupportedOperationException();
-    }
+        @Override
+        public Iterator<Student> iterator() {
+            return new Iterator<Student>() {
+                final Iterator<Node> iterator = new TreeIterator(root);
+
+                @Override
+                public boolean hasNext() {
+                    return iterator.hasNext();
+                }
+
+                @Override
+                public Student next() {
+                    return iterator.next().student;
+                }
+            };
+        }
+
+        private class TreeIterator implements Iterator<Node>{
+            private Node next;
+
+            private TreeIterator(Node root){
+                next = root;
+                goLeft();
+            }
+
+            private void goLeft() {
```

Comment:
maybe something like
```suggestion
            private void goLeftMost() {
```